common/transport/amqp/devdoc/sender_link_requirements.md

```diff
@@ -0,0 +1,88 @@
+
+### attach(callback: (err?: Error) => void): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_004: [** The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `SenderLink` instance. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_022: [** The `attach` method shall call the `callback` if the link was successfully attached. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_005: [** If the `amqp10.AmqpClient` emits an `errorReceived` event during the time the link is attached, the `callback` function shall be called with this error. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_006: [** The `SenderLink` object should subscribe to the `detached` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_007: [** The `SenderLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_008: [** If the `amqp10.AmqpClient` fails to create the link the `callback` function shall be called with this error object. **]**
+
+### detach(): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_009: [** The `detach` method shall detach the link created by the `amqp10.AmqpClient` underlying object. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_017: [** The `detach` method shall return the state machine to the `detached` state. **]**
+
+### send(message: AmqpMessage, callback: (err?: Error, result?: results.MessageEnqueued) => void): void
+
+**SRS_NODE_AMQP_SENDER_LINK_16_010: [** The `send` method shall use the link created by the underlying `amqp10.AmqpClient` to send the specified `message` to the IoT hub. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_011: [** If the state machine is not in the `attached` state, the `SenderLink` object shall attach the link first and then send the message. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_012: [** If the message cannot be sent the `callback` shall be called with an `Error` object describing the AMQP error reported by the service. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_013: [** If the message is successfully sent, the `callback` shall be called with a first parameter (error) set to `null` and a second parameter of type `MessageEnqueued`. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_014: [** If the link is detached while a message is being sent, the `callback` shall be called with an `Error` object describing the AMQP error that caused the detach to happen in the first place. **]**
+
+### Events
+
+**SRS_NODE_AMQP_SENDER_LINK_16_015: [** If a `detached` or `errorReceived` event is emitted by the `ampq10` link object, the `SenderLink` object shall return to the `detached` state. **]**
+
+### Internal state machine
+
+**SRS_NODE_AMQP_SENDER_LINK_16_016: [** If returning to the `detached` state because of an error that didn't happen while trying to attach the link or send a message, the sender link shall call emit an `error` event with that error. **]**
```

Comment:
the amqp10 folks came up with this idea and I remember a discussion with them telling me they regret it. it's our job to subscribe to the error event on the links, not the user's. to a certain extent, the link object itself shouldn't even be exposed to the user. therefore, i have no problem emitting the standard error event and differentiating based on type or error code. The client will handle that error (with a retry for example) and decide if it is necessary to bubble it up to the user or not.


---
In reply to: [128120964](https://github.com/Azure/azure-iot-sdk-node/pull/55#discussion_r128120964) [](ancestors = 128120964)